### ISSUE TITLE
Data-cells containing only replaced elements inappropriately flagged as not being data-cells

### ISSUE THREAD

**Issue by JaredNoble** [dequelabs/axe-core#289]
In the context of a simple two-column table that otherwise generates no accessibility warnings, the following row will cause AXE to generate a warning:
` <tr>`
`    <th><label for="Name">Name</label></th>`
`    <td><input type="text" id="Name" name="Name" size="40"></td>`
`  </tr>`

"**Axe Error**: All th element and elements with role=columnheader/rowheader must data cells which it describes" _[errors in the warning text are part of the product - but not my point]_.

In my testing, ANY cell that contains _only_ whitespace and/or replaced elements (img, input, etc) is treated as NOT being a data-cell. Since in this test case there is only one &lt;td> associated with the &lt;th>, the error is displayed. 
Including any non-whitespace, non-replaced element (such as *) will prevent the error from being displayed.

It appears that the validation algorithm discards all replaced elements in the td, and if the resulting cell is either empty or contains only whitespace, it is considered to not be a data-cell.  I believe this is a bug, and have not been able to find justification in WCAG for this interpretation. I freely admit I may have overlooked a rule, but would like clarification as we are using AXE to verify changes to our pages based on a Deque site assessment.

1. What is the justification for assuming that replaced elements are not content in a data cell?
2. What is the justification for declaring that an empty td is not a data-cell?

**Comment by marcysutton** [comment]
I'm not convinced that a `th` containing a label and a `td` containing an input is actually a data table. Isn't that a layout table? Is there more to the table markup than what's shown in your snippet above? 

I ask, because browsers have algorithms to determine whether tables are actually data tables or presentational–which the aXe rules try to follow by saying a table with more than 4 columns (among other checks) is a layout table. You can inspect this logic yourself in [/lib/commons/table/is-data-table.js](https://github.com/dequelabs/axe-core/blob/master/lib/commons/table/is-data-table.js).

**Comment by JaredNoble** [comment]
There is of course more to the table that the snippet provided, and I understand your question / concern about whether or not this is a data or a layout table. In _this_ example the table is part of a vertical structured record presentation, which currently has only one record displayed. In other circumstances there may be multiple records displayed side-by-side (think product comparison grid).

At any rate, I think this is orthogonal to the issue, which consists of 2 questions:
1. Why assume that replaced elements in a td equate to an empty td?
2. Why declare that an empty td is not a data-cell? As far as I can tell ARIA rules only specify not having empty th elements, not td.

**Comment by marcysutton** [comment]
> In *this* example the table is part of a vertical structured record presentation, which currently has only one record displayed.

Is this handled in the raw markup or with CSS?

Can you provide a more complete table structure so we can debug it? An HTML demo file or Codepen/Jsbin/etc. would be very helpful.

**Comment by marcysutton** [comment]
@JaredNoble I believe we have fixed this internally, but to verify I need a demo from you. Is that something you can provide?

**Comment by JaredNoble** [comment]
Sorry for the extended delay. I've been away for a while.
The following codepen shows the error. Obviously aXe also has a few things to say about the codepen code as well, but you will see the error regarding the last th element in the table, despite the fact that the column contains 3 data cells in the following rows. One contains an image, one contains a form element, and the last contains a non-breaking space. The table contains 28 total cells (7 scoped headers, and 3 rows of 7 data cells.)

http://codepen.io/jnoble/pen/peQPaw

**Comment by JaredNoble** [comment]
On a related note, if I have this same table as the only element in a standalone html5 page, I get the same error. If I instead place the table in a page with an html4 doctype, I get a new error: "scope attribute should be used correctly". The Error summary shows the following:
 Fix all of the following:
    In HTML 5, scope attributes may only be used on table header elements (<th>)

So apparently there is something else with html4 pages that triggers complaints about html5 restrictions, despite the fact that the page itself is completely compliant with this html5 requirement. Odd.

**Comment by JaredNoble** [comment]
Marcy, I don't have a public publishing environment that I can host arbitrary demo code on, and codepen does not allow me to control the doctype and present just a bare HTML4 page. I'm not sure how I can provide a demo for this "related note" comment.

**Comment by marcysutton** [comment]
@JaredNoble it's ok, we can play with it. Thanks for the demos! Stay tuned.

**Comment by marcysutton** [comment]
I'm not able to reproduce this issue anymore when I run aXe against that Codepen page, so I'm going to close it. Let us know if you're still having issues. https://s.codepen.io/jnoble/debug/peQPaw/vWMRwnOwaGnr
